starbuck main stage (#19725)
supabase/supabase: apps/www/_blog/2023-12-14-supabase-auth-dentity-linking-session-control.mdx

---

> title: 'Supabase Auth: Identity Linking, Session Control, Passwords and Hooks'

Reviewer: ```suggestion
title: 'Supabase Auth: Identity Linking, Hooks, and Have I Been Pwned integration'
```

> title: 'Supabase Auth: Identity Linking, Hooks, and Have I Been Pwned integration'

Reviewer 2: ```suggestion
title: 'Supabase Auth: Identity Linking, Hooks, and HaveIBeenPwned integration'
```

> description: Supacharging Supabase Auth with Identity Linking, Session Control, Passwords, and Hooks

Reviewer: ```suggestion
description: Four major Auth features: Identity Linking, Session Control, Leaked Password Protection, and Hooks
```

(let's avoid using `supa` wherever possible)

launchweek: X
categories:
  - product
tags:
  - launch-week
  - auth
date: '2023-12-14'
toc_depth: 3
author: stojan,joel,kangmingtay
image: lwx-supabase-auth/auth-og.png
thumb: lwx-supabase-auth/auth-thumb.png
---

> Welcome to Launch Week X Day 4! We're excited to announce four new features for Supabase Auth:

Reviewer: ```suggestion
We're excited to announce four new features for Supabase Auth:
```

1. Identity Linking
2. Session Control
3. Leaked Password Protection
4. Auth Hooks with Postgres functions

<div className="video-container">
  <iframe
    className="w-full"
    src="https://www.youtube-nocookie.com/embed/LF8GABnAFyE"
    title="YouTube video player"
    allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
    allowfullscreen
  />
</div>
> Watch a demonstration of some of the new features

## Identity Linking

> When a user signs in, an identity is created with the method and sign-in provider. Historically, [Supabase Auth](https://supabase.com/docs/guides/auth) has been automatically linking identities to a user if the identity shares the same verified email as the user. This is convenient to de-duplicate user accounts. However, some developers also need the flexibility to link accounts that don’t share the same email. 

Reviewer 2: ```suggestion
When a user signs in, an identity is created with the authentication method and sign-in provider. Historically, [Supabase Auth](https://supabase.com/docs/guides/auth) has been automatically linking identities to a user if the identity shares the same verified email as the user. This is convenient to de-duplicate user accounts. However, some developers also need the flexibility to link accounts that don’t share the same email. 
```

IMAGE GOES HERE

> For more information about how identity linking works, see the [Identity Linking docs](https://supabase.com/docs/guides/auth/auth-identity-linking).

## Session Control

Supabase Auth manages the full session lifecycle from the moment your user signs into your application. This involves the following steps:

1. Creating the session for the user.
2. Refreshing the session to keep it active.
3. Revoking the session upon expiry or logout. 

For developers who want finer control over their users’ sessions, we have exposed 3 new settings:

- **Time-box user sessions**: Force users to sign in again after a time interval.
- **Inactivity Timeout**: Force users to sign in again if they’re inactive for a time interval.
- **Single session per user**: Restrict users to a single session. The most recently active session is kept, and all others are terminated.

These session control settings are available on the Pro plan and above.

IMAGE GOES HERE

> For more information, see the [Session Management docs](https://supabase.com/docs/guides/auth/sessions).

## Leaked Password Protection

> Passwords often lack security due to user habits like weak choices or reuse across several sites. 

Reviewer 2: hmm the phrasing sounds strange to me, i prefer:
```suggestion
Passwords can be inherently insecure due to common user behaviors like choosing guessable passwords or reusing them across different platforms. 
```

- Apply custom rules for password validation attempts

We aim to maintain a straightforward and seamless Supabase Auth experience. It should work effortlessly for most developers, requiring no customization. However, recognizing the diversity of apps, you can now extend standard Auth features through Auth Hooks.

Auth Hooks are simply Postgres functions that run synchronously at key points in the Auth lifecycle, to change the outcome of the action. 

For example, to customize the JWT claims with Auth Hooks, you can create a Postgres function that accepts the JWT claims in the first argument and returns the JWT you wish to be used by Supabase Auth. 

Suppose you’re creating a gamified application and you wish to attach the user’s level to the JWT as a custom claim:

```sql
create function custom_access_token_hook(event jsonb)
returns jsonb
language plpgsql
as $$
declare
  user_level jsonb;
begin
  -- fetch the current user's level
  select 
    to_jsonb(level) into user_level 
  from profiles 
  where 
    user_id = event->>'user_id'::uuid;

  -- change the event.claims.level
  return jsonb_set(
		event, 
		'{claims,level}', 
		user_level);

end;
$$
```

> Then once you’ve created the function in the database, you only need to register it with Supabase Auth:

Reviewer 2: ```suggestion
Once you’ve created the function in the database, you only need to register it with Supabase Auth:
```

IMAGE GOES HERE

Initially, you can register an Auth Hook for the following points in the flow:

- **Custom access token:** called each time a new JWT is generated.

> - **MFA verification attempt**: called each time an MFA factor is verified, allowing you finer control over detecting and blocking attempts.

Reviewer 2: ```suggestion
- **MFA verification attempt**: called each time an MFA factor is verified, allowing finer control over detecting and blocking attempts.
```

> - **Password verification attempt**: called each time a password is used to sign-in a user, allowing you finer control over the security of the user’s accounts.

Reviewer 2: ```suggestion
- **Password verification attempt**: called each time a password is used to sign-in a user, allowing finer control over the security of the user’s accounts.
```

> Auth Hooks is available today for self-hosting and will be rolling out to the platform next month. Reach out to us via [support](https://supabase.help) if you need access sooner!

Reviewer 2: ```suggestion
Auth Hooks is available today for self-hosting and will be rolled out to the platform next month. Reach out to us via [support](https://supabase.help) if you need access sooner!
```
